snowcap/api/lua/snowcap/layer.lua

```diff
@@ -130,28 +138,37 @@ function layer.new_widget(args)
         layer_id = layer_id,
     }, function(response)
         local widget_id = response.widget_id or 0
+        local msg = nil
+
         if response.button then
-            if callbacks[widget_id] then
-                args.program:update(callbacks[widget_id])
-                local widget_def = args.program:view()
-                callbacks = {}
-
-                require("snowcap.widget")._traverse_widget_tree(
-                    widget_def,
-                    callbacks,
-                    function(callbacks, widget)
-                        if widget.button and widget.button.on_press then
-                            callbacks[widget.button.widget_id] = widget.button.on_press
-                        end
-                    end
+            msg = callbacks[widget_id]
+        elseif response.mouse_area then
+            widget_id = response.mouse_area.unique_id or widget_id
```

Comment:
Here, having the `unique_id` + `widget_id` unified `message` would mage it less weird, since you wouldn't need to first check what kind of message the response contain it is before using it.

As a side note:
It's possible to define `callbacks` to accept a table as key and do the right thing by setting `__index` and `__newindex` as functions in its metatable.

 However, I think it's ok to just extract the 'correct one' depending on whether the `unique_id` is set or not, and I don't really like to do that kind of `metatable` magic unless absolutely needed.